e2e/connect/acls.go

```diff
@@ -0,0 +1,401 @@
+	}, &capi.WriteOptions{Token: tc.consulMasterToken})
+	r.NoError(err, "failed to create operator token")
+	tc.consulTokenIDs = append(tc.consulTokenIDs, token.AccessorID)
+	return token.SecretID
+}
+
+func (tc *ConnectACLsE2ETest) TestConnectACLsRegister_MasterToken(f *framework.F) {
+	t := f.T()
+	r := require.New(t)
+
+	t.Log("test register Connect job w/ ACLs enabled w/ master token")
+
+	jobID := "connect" + uuid.Generate()[0:8]
+	tc.jobIDs = append(tc.jobIDs, jobID)
+
+	jobAPI := tc.Nomad().Jobs()
+
+	job, err := jobspec.ParseFile(demoConnectJob)
+	r.NoError(err)
+
+	// Set the job file to use the consul master token.
+	// One should never do this in practice, but, it should work.
+	// https://www.consul.io/docs/acl/acl-system.html#builtin-tokens
+	job.ConsulToken = &tc.consulMasterToken
+
+	resp, _, err := jobAPI.Register(job, nil)
+	r.NoError(err)
+	r.NotNil(resp)
+	r.Zero(resp.Warnings)
+}
+
+func (tc *ConnectACLsE2ETest) TestConnectACLsRegister_MissingOperatorToken(f *framework.F) {
+	t := f.T()
+	r := require.New(t)
+
+	t.Log("test register Connect job w/ ACLs enabled w/o operator token")
+
+	job, err := jobspec.ParseFile(demoConnectJob)
+	r.NoError(err)
+
+	jobAPI := tc.Nomad().Jobs()
+
+	// Explicitly show the ConsulToken is not set
+	job.ConsulToken = nil
+
+	_, _, err = jobAPI.Register(job, nil)
+	r.Error(err)
+
+	t.Log("job correctly rejected, with error:", err)
+}
+
+func (tc *ConnectACLsE2ETest) TestConnectACLsRegister_FakeOperatorToken(f *framework.F) {
+	t := f.T()
+	r := require.New(t)
+
+	t.Log("test register Connect job w/ ACLs enabled w/ operator token")
+
+	policyID := tc.createConsulPolicy(consulPolicy{
+		Name:  "nomad-operator-policy",
+		Rules: `service "count-api" { policy = "write" } service "count-dashboard" { policy = "write" }`,
+	}, f)
+	t.Log("created operator policy:", policyID)
+
+	// generate a fake consul token token
+	fakeToken := uuid.Generate()
+	job := tc.parseJobSpecFile(t, demoConnectJob)
+
+	jobAPI := tc.Nomad().Jobs()
+
+	// deliberately set the fake Consul token
+	job.ConsulToken = &fakeToken
+
+	// should fail, because the token is fake
+	_, _, err := jobAPI.Register(job, nil)
+	r.Error(err)
+	t.Log("job correctly rejected, with error:", err)
+}
+
+func (tc *ConnectACLsE2ETest) TestConnectACLs_ConnectDemo(f *framework.F) {
+	t := f.T()
+	r := require.New(t)
+
+	t.Log("test register Connect job w/ ACLs enabled w/ operator token")
+
+	// === Setup ACL policy and token ===
+
+	// create a policy allowing writes of services "count-api" and "count-dashboard"
+	policyID := tc.createConsulPolicy(consulPolicy{
+		Name:  "nomad-operator-policy",
+		Rules: `service "count-api" { policy = "write" } service "count-dashboard" { policy = "write" }`,
+	}, f)
+	t.Log("created operator policy:", policyID)
+
+	// create a Consul "operator token" blessed with the above policy
+	operatorToken := tc.createOperatorToken(policyID, f)
+	t.Log("created operator token:", operatorToken)
+
+	// === Register the Nomad job ===
```

Comment:
The logic you have here looks solid, but it looks like it could be generalized into `e2eutil/utils.go` to make what's already there more robust? It looks from your commit like what we have with `require.Eventually` there is brittle, so if we could reuse this that would be an across-the-board improvement.